app/middlwares/Auth0Middleware.go

```diff
@@ -0,0 +1,75 @@
+package middlwares
+
+import (
+	"context"
+	jwtmiddleware "github.com/auth0/go-jwt-middleware/v2"
+	"github.com/gin-gonic/gin"
+	adapter "github.com/gwatts/gin-adapter"
+
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/auth0/go-jwt-middleware/v2/jwks"
+	"github.com/auth0/go-jwt-middleware/v2/validator"
+)
+
+type a0Middleware struct{}
+
+// NewAuth0Middleware is a constructor
+func NewAuth0Middleware() Middleware {
+	return &a0Middleware{}
+}
+
+// CustomClaims contains custom data we want from the token.
+type CustomClaims struct {
+	Scope string `json:"scope"`
+}
+
+// Validate does nothing for this example, but we need
+// it to satisfy validator.CustomClaims interface.
+func (c CustomClaims) Validate(ctx context.Context) error {
+	return nil
+}
+
+// AuthMiddleware checks if it has a jwt token and then requests the auth service to verify it for us
+func (h *a0Middleware) AuthMiddleware() gin.HandlerFunc {
+	issuerURL, err := url.Parse("https://" + os.Getenv("AUTH0_DOMAIN") + "/")
+	if err != nil {
+		log.Fatalf("Failed to parse the issuer url: %v", err)
+	}
+
+	provider := jwks.NewCachingProvider(issuerURL, 5*time.Minute)
+
+	jwtValidator, err := validator.New(
+		provider.KeyFunc,
+		validator.RS256,
+		issuerURL.String(),
+		[]string{os.Getenv("AUTH0_AUDIENCE")},
+		validator.WithCustomClaims(
+			func() validator.CustomClaims {
+				return &CustomClaims{}
+			},
+		),
+		validator.WithAllowedClockSkew(time.Minute),
+	)
+	if err != nil {
+		log.Fatalf("Failed to set up the jwt validator")
+	}
+	errorHandler := func(w http.ResponseWriter, r *http.Request, err error) {
+		log.Printf("Encountered error while validating JWT: %v", err)
+
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusUnauthorized)
+		w.Write([]byte(`{"message":"Failed to validate JWT."}`))
+	}
+
+	middleware := jwtmiddleware.New(
+		jwtValidator.ValidateToken,
+		jwtmiddleware.WithErrorHandler(errorHandler),
+	)
+
+	return adapter.Wrap(middleware.CheckJWT)
```

Comment:
why do you use an adapter?

Reply:
to convert golang net/http middleware to gin middleware :)